File: guessingcircle/submissions/time_limit_exceeded/per_n3_v1.cc
#include<bits/stdc++.h>
using namespace std;

typedef unsigned long long word;

int main(void) {
    int n;
    scanf("%d", &n);
    vector<int> circle(n);
    unordered_map<int, int> num_idx;
    vector<vector<int>> positions;
    for (int i = 0; i < n; ++i) {
        scanf("%d", &circle[i]);
        if (num_idx.find(circle[i]) == num_idx.end()) {
            num_idx[circle[i]] = num_idx.size();
            positions.push_back({});
        }
        positions[num_idx[circle[i]]].push_back(i);
    }
    int m = num_idx.size();
    int words = (m+63)/64;
    vector<vector<word>> possL(m, vector<word>(words, 0));
    vector<vector<word>> possR(m, vector<word>(words, 0));
    for (int a = 0; a < m; ++a)
        for (int b = 0; b < m; ++b)
            for (int i: positions[a])
                for (int j: positions[b]) {
                    int dR = (j-i+n) % n, dL = (i-j+n) % n;
                    if (dL <= dR)
                        possL[a][b/64] |= word(1) << (b % 64);
                    if (dR <= dL)
                        possR[a][b/64] |= word(1) << (b % 64);
                }
    vector<int> ans;
    for (auto p: num_idx) {
        auto &[num, idx] = p;
        bool ok = true;

        for (int j = 0; ok && j < m; ++j) {
            if (j == idx) continue;
            bool same = true;
            for (int k = 0; same && k < words; ++k)
                same &= ((possL[idx][k] & possL[j][k]) | (possR[idx][k] & possR[j][k])) == (possL[idx][k] | possR[idx][k]);
            ok &= !same;
        }
        if (ok) ans.push_back(num);
    }
    if (ans.empty())
        printf("none\n");
    else {
        sort(ans.begin(), ans.end());
        for (int x: ans)
            printf("%d\n", x);
    }
}

File: guessingcircle/submissions/time_limit_exceeded/per_n3_v1.py
#!/usr/bin/env python3
import sys

BITS = 62
ALL = (1<<BITS)-1

n = int(input())
circle = list(map(int, input().split()))

num_idx = {}
positions = []
for (i, c) in enumerate(circle):
    if c not in num_idx:
        num_idx[c] = len(num_idx)
        positions.append([])
    positions[num_idx[c]].append(i)

m = len(num_idx)
fill = ALL ^ ((1<<(m%BITS)) - 1) if m % BITS else 0
words = (m+BITS-1)//BITS
possL = [[0]*words for _ in range(m)]
possR = [[0]*words for _ in range(m)]
for a in range(m):
    possL[a][words-1] |= fill
    possR[a][words-1] |= fill
    for b in range(m):
        for i in positions[a]:
            for j in positions[b]:
                dR = (j-i) % n
                dL = (i-j) % n
                if dL <= dR:
                    possL[a][b//BITS] |= 1<<(b%BITS)
                if dR <= dL:
                    possR[a][b//BITS] |= 1<<(b%BITS)


ans = []
for (num, idx) in num_idx.items():
    ok = 1
    L1 = possL[idx]
    R1 = possR[idx]
    for j in range(m):
        if j == idx: continue
        same = 1
        L2 = possL[j]
        R2 = possR[j]
        for k in range(words):
            same &= ((L1[k] & L2[k]) | (R1[k] & R2[k])) == ALL
            if not same: break
        ok &= not same
        if not ok: break
    if ok: ans.append(num);

if not ans:
    print("none");
else:
    ans.sort()
    print(*ans)
